### DISCUSSION TITLE
Allow passing of DAX expression via command line

### DISCUSSION THREAD

**Discussion by NickEPalmer** [DaxStudio/DaxStudio#519] [Ideas]
This question sort of piggy backs on an earlier question asked.  Thru some digging I found that I can pass an AS server and database to the DaxStudio via the command line and have DaxStudio come up connected to the that AS server and database which is really cool.

In our app we allow the user to edit DAX expressions but don't have anywhere near the editor that DaxStudio studio has for editing DAX.  

Would it be possible to add a way to pass a DAX expression in addition to the AS server and database to DaxStudio via the command line?  And as a bonus, allow us some way to pass that back? The user can always copy/paste the completed DAX from DaxStudio back into our application but if removing that step would remove any chance of user error.

Thanks
Nick

**Comment by dgosbell** [comment]
> This question sort of piggy backs on an earlier question asked. Thru some digging I found that I can pass an AS server and database to the DaxStudio via the command line and have DaxStudio come up connected to the that AS server and database which is really cool.

Yes, this is how the External Tools feature in Power BI passes the connection details to DAX Studio

> Would it be possible to add a way to pass a DAX expression in addition to the AS server and database to DaxStudio via the command line?

It probably would not be too hard to do this, but I'm not sure how we would pass the text back. 

Could you use a temporary file as the interface? If you wrote out the expression to a temp file, then started DAX Studio with a command line like the following:

```DaxStudio.exe /server <myserver> /database <database> /file <path to file>``` 

Then you could either watch the temp file for changes or if you started DaxStudio.exe using `start /wait` this should block your call until DAX Studio was closed. 

We already have a `/file`  parameter, but currently we don't currently support it's use in combination with `/server` and `/database` but that would not be hard to add.

**Reply by NickEPalmer** [reply]
Thanks for response.

What you are proposing would totally work.  We are currently launching DAX Studio the same way that Power BI is and adding the ability to pass a file with the DAX query would be great.  

I'm assuming we'd have to pass a .DAX file correct?  From what I can tell from looking at a .DAX file in notepad creating a temp DAX file wouldn't be that hard.

**Comment by dgosbell** [comment]
Yes, passing a .dax file would be the safest approach, but they are just utf-8 text files. I'll look at making the server, database, file parameter combination work then, that should be pretty simple.

**Reply by NickEPalmer** [reply]
Sounds perfect! Can't wait :)

**Reply by NickEPalmer** [reply]
Hi. Just curious if you guys have had a chance to implement this change? Should I add this as a bug/issue?
Thanks

**Comment by dtate22** [comment]
On this same thread of thinking...
- I would like to Write Unit Tests with DAX Studio?
- Is there any plan to do that?
      - since obviously, I can provide parameters to DAX

**Reply by dgosbell** [reply]
@dtate22 so this has come up as a suggestion before (see #368) the biggest limitation is that DAX Studio can really only test the DAX layer of a BI solution and you would need other things to ensure that the ETL is also correct. It would probably also benefit from being able to be run from the command line which is another suggestion that we are considering (see #92) 

I have some ideas in mind about how this might be implemented including things like being able to assert against a scalar value or being able to assert against values in a csv file, but these are all just concepts at the moment.

I'm curious about your mention of providing parameters. How do you see parameters being used in DAX unit tests? Can you provide some examples of what you are thinking about in this regard?
